src/iter/walk_tree.rs

```diff
@@ -0,0 +1,545 @@
+use crate::iter::plumbing::{bridge_unindexed, Folder, UnindexedConsumer, UnindexedProducer};
+use crate::prelude::*;
+use std::iter::once;
+use std::marker::PhantomData;
+
+#[derive(Debug)]
+struct WalkTreePrefixProducer<'b, S, B, I> {
+    to_explore: Vec<S>, // nodes (and subtrees) we have to process
+    seen: Vec<S>,       // nodes which have already been explored
+    breed: &'b B,       // function generating children
+    phantom: PhantomData<I>,
+}
+
+impl<'b, S, B, I, IT> UnindexedProducer for WalkTreePrefixProducer<'b, S, B, I>
+where
+    S: Send,
+    B: Fn(&S) -> I + Send + Sync,
+    IT: DoubleEndedIterator<Item = S>,
+    I: IntoIterator<Item = S, IntoIter = IT> + Send,
+{
+    type Item = S;
+    fn split(mut self) -> (Self, Option<Self>) {
+        // explore while front is of size one.
+        while self.to_explore.len() == 1 {
+            let front_node = self.to_explore.pop().unwrap();
+            self.to_explore
+                .extend((self.breed)(&front_node).into_iter().rev());
+            self.seen.push(front_node);
+        }
+        // now take half of the front.
+        let right_children = split_vec(&mut self.to_explore);
+        let right = right_children
+            .map(|mut c| {
+                std::mem::swap(&mut c, &mut self.to_explore);
+                WalkTreePrefixProducer {
+                    to_explore: c,
+                    seen: Vec::new(),
+                    breed: self.breed,
+                    phantom: PhantomData,
+                }
+            })
+            .or_else(|| {
+                // we can still try to divide 'seen'
+                let right_seen = split_vec(&mut self.seen);
+                right_seen.map(|s| WalkTreePrefixProducer {
+                    to_explore: Default::default(),
+                    seen: s,
+                    breed: self.breed,
+                    phantom: PhantomData,
+                })
+            });
+        (self, right)
+    }
+    fn fold_with<F>(mut self, mut folder: F) -> F
+    where
+        F: Folder<Self::Item>,
+    {
+        // start by consuming everything seen
+        for s in self.seen {
+            folder = folder.consume(s);
```

Comment:
This can use `folder.consume_iter(self.seen)` instead of a manual loop, and one `full()` check afterward.